### PR TITLE
Support extended protocol and prepared statements

### LINKED ISSUE #118: Error with extended query protocol
```
postgres=# create table tbl(i int4);
CREATE TABLE
postgres=# select * from tbl  \bind \g
ERROR:  unrecognized node type: 1189775888
```

See https://github.com/duckdb/pg_duckdb/blob/177b12b7a39fc14e9c0ffb136c8c9c957dc3f2d7/src/pgduckdb_planner.cpp#L91

The problem is that the type stored in `custom_private` is supposed to be a copyable `Node` type, which the duckdb connection and prepared query types are not.

### PR DESCRIPTION
**Pull request by JelteF**
Previously prepared statements and the extended protocol were broken. We need to support this feature because pretty much every client library uses it. The reason that prepared statements did not work is that for prepared statements to work, a Postgres plan needs to be fully copyable using Postgres functionality. We were previously storing the DuckDB plan in the Postgres plan, but since Postgres did not know how to copy the DuckDB plan it would fail.

To work around that problem we now store the full Postgres `Query` structure inside the DuckDB CustomScan node. Then during execution we use this `Query` structure to re-plan the query during execution. This works but it has the big downside that we're now planning the query twice in DuckDB, once during planning to get the the column types of the result and once during execution. For now that seems acceptable, so we can at least support prepared statements and the extended protocol. In the future we might want to do something smarter though, like serializing and deserializing the DuckDB plan that is created at planning time.

Note: This does not support all our supported types as arguments to a prepared statement yet. After merging I'll open separate issue to track adding the rest of these types. 

Fixes #118

### CODE REVIEW

File: src/pgduckdb_types.cpp
Comment (JelteF) [inline src/pgduckdb_types.cpp:735]: As you can see there's a few types that we support in other places that I didn't implement here yet. Also if you look at the added tests you'll see that there are a bunch of issues with floats and string arguments for some reason not working as expected in comparisons. 

In my opinion it's important to support them and fix those errors, but I don't think it necessarily needs to be part of this PR. This PR is already a bit larger than I'd like it to be. So I'd rather merge this soonish if we're happy with the general design, and then fix the remaining issues in follow up PRs. Especially since making DuckDB a singleton is a relatively large change in design that I feel would benefit from early use.

File: src/pgduckdb_planner.cpp
Comment (JelteF) [inline src/pgduckdb_planner.cpp:59]: Putting an extra note here for visibility. This is very important to do, there can be big bugs otherwise. But I'd prefer to do this after merging #143, because that makes it easy to add tests for explain.

EXPLAIN in general is broken btw, if you use it on a extended protocol statement with parameters. That should be fixed too. The reason for that is that `Duckdb_ExplainCustomScan` doesn't provide the parameters to the duckdb `Execute` function yet. Basically we need to do there what I've now only done in `ExecuteQuery`, which is called by `Duckdb_ExecCustomScan`.

File: src/pgduckdb_duckdb.cpp
Comment (Y--) [reply src/pgduckdb_duckdb.cpp]: Yeah it was clumsy anyway, I didn't like how we were storing anything "per-connection" in the global config. I meant to rework it when working on the proper "singleton" implementation, so I did it here: [`72232cc` (#147)](https://github.com/duckdb/pg_duckdb/pull/147/commits/72232cc46dab51069d39076f8d4e2e5e6a9cb98b). Basically we're storing in the `ClientContext` under a well-defined key.

Let me know what you think.

File: src/pgduckdb_node.cpp
Comment (Y--) [inline src/pgduckdb_node.cpp:93]: minor nits (constify upper bound, prefix increment and hoist loop variables):
```suggestion
		const auto n = pg_params->numParams;
		ParamExternData *pg_param = NULL;
		ParamExternData tmp_workspace;
		for (int i = 0; i < n; ++i) {
```

File: src/pgduckdb_node.cpp
Comment (Y--) [inline src/pgduckdb_node.cpp:114]:
```suggestion
		elog(ERROR, "DuckDB execute returned an error: %s", pending->GetError().c_str());
```
Also, I wonder if we should surface the "re-planning" vs "execute" details to end users?

File: src/pgduckdb_types.cpp
Comment (Y--) [inline src/pgduckdb_types.cpp:722]: Don't we need the whole de-toastification logic here?

**Review by Y-- — APPROVED** [review]
Overall LGTM, thanks for picking up! Would like to hear what other thinks too cc @Tishj @mkaruza

File: src/pgduckdb_duckdb.cpp
Comment (mkaruza) [inline src/pgduckdb_duckdb.cpp:147]: This will add multiple `PostgresReplacementScan` into `replacement_scans` vector. Maybe add this as part of singleton construction?

File: src/pgduckdb_types.cpp
Comment (mkaruza) [reply src/pgduckdb_types.cpp:735]: This function is only used for binding? Maybe rename it, also we don't need to have commented out what is not supported (probably ticket should be created at this point).

**Review by mkaruza — CHANGES_REQUESTED** [review]
Personally i would like to see this PR split int two parts. One is which enabled singleton duckdb object and another that deals with binding.

File: src/pgduckdb_duckdb.cpp
Comment (JelteF) [reply src/pgduckdb_duckdb.cpp]: I think that approach makes sense. On request from @mkaruza I moved the singleton logic into its own PR in #174. The only change I made is that the replacement scan now gets added to the config on Database creation, not on connection creation. To make this succeed, I made the replacement scan return `nullptr` when the `"postgres_scan"` key was not found.

File: src/pgduckdb_types.cpp
Comment (JelteF) [reply src/pgduckdb_types.cpp:722]: Toast is only used when storing values on disk. Since this is a parameter passed in by the client we don't need it here (assuming we only ever pass parameter values to the function). I added this consideration to the function comment now.

File: test/pycheck/explain_test.py
Comment (JelteF) [inline test/pycheck/explain_test.py:5]: I needed to change this definition to not create an index because index scans are currently broken, see #183 for details.

File: src/pgduckdb_planner.cpp
Comment (JelteF) [reply src/pgduckdb_planner.cpp]: EXPLAIN is now fixed. And I was wrong about it being necessary to move this to custom_private to avoid bugs. I added a comment in the `DuckdbExplainOneQueryHook` elaborating on why the current design is fine.

File: src/pgduckdb_types.cpp
Comment (JelteF) [reply src/pgduckdb_types.cpp]: I renamed the function a bit and removed the commented out code. After we merge this I'll create a ticket to track support for the other types.

**Review by mkaruza — APPROVED** [review]

**Pull request merged**
